src/core/Akka/Util/ByteString.cs

```diff
@@ -580,13 +580,12 @@ public IEnumerator<byte> GetEnumerator()
 
         public string ToString(Encoding encoding)
         {
-            var builder = new StringBuilder(_count);
-            foreach (var buffer in _buffers)
-            {
-                var part = encoding.GetString(buffer.Array, buffer.Offset, buffer.Count);
-                builder.Append(part);
-            }
-            return builder.ToString();
+            if (IsCompact)
+                return encoding.GetString(_buffers[0].Array, _buffers[0].Offset, _buffers[0].Count);
+
+            byte[] buffer = ToArray();
```

Comment:
What advantage does that give over using `StringBuilder`? I'm just curious.

Reply:
Hi! The problem is that when you have a `ByteString` composed of several `ByteString`s and some of those `ByteStrings` represent incomplete Unicode sequences (e.g. `"AB"` in Unicode is 4 bytes long, and if you split the `ByteString` after 3 bytes, resulting in a `ByteString` representing 1-and-a-half characters), then the resulting string will be incorrect because some of its components will be incorrect.

Reply:
Most of the existing code and tests used UTF-8, which encodes characters from the ASCII character set using only a single byte (so the issue may not have come up before). As soon as I switched to Unicode it gave wrong results unless I called `.Compact()` before calling `.ToString()`.

Reply:
[Example](https://github.com/akkadotnet/akka.net/pull/3148/files#diff-07edeee541b3c2f1fbe2a9aab5458214R135)